### ISSUE TITLE
Bumping d3 to v4.0.0

### ISSUE THREAD

**Issue by etpinard** [plotly/plotly.js#424]
*A discussion on what modules we need in [d3 v4.0.0](https://github.com/d3).*

**Comment by monfera** [comment]
@etpinard these are probably well known to all of us, so it's a 'just in case':

- It's currently in alpha, so it's good to think about it, but I'm not aware of a target release date which makes it hard to incorporate in a roadmap
- feels like an overall refactoring and API improvements for an incrementally better D3, rather than a revolution or leap in some direction; some favorites are better support for non-SVG (Canvas, WebGL) targets, better timer, transitions and of course the Matplotlib [Viridis](https://github.com/d3/d3-scale#scaleViridis) palettes
- .. but a real deep article from Bostock seems to be https://medium.com/@mbostock/what-makes-software-good-943557f8a488#.9lyi5ovkg
- while inclusion of D3 4.0 can be done via ES6 modules, it does not mean that payload size would be reduced: because of the method chaining style of D3, if you need the smallest thing from e.g. d3-scale, it'll pull in the entirety of d3-scale
- these are the [main options](https://docs.google.com/presentation/d/1piFaL4ttKNAv6_lkUgJLLSIRSah7blvd8ghqMFEeNnk/edit#slide=id.g100bd858fa_0_78 ) of including D3 4.0 in the project- personally I'd go for 
- the improvements in D3 4.0 often mean that the code [needs changing](https://docs.google.com/presentation/d/1piFaL4ttKNAv6_lkUgJLLSIRSah7blvd8ghqMFEeNnk/edit#slide=id.g100bd858fa_0_167), so migration of a large project takes some time

**Comment by n-riesco** [comment]
@monfera Thx for the links!

**Comment by monfera** [comment]
Ann. by Mike Bostock: [D3 4.0 is now in `master` and a release is coming up in a month or two](https://twitter.com/mbostock/status/731159572568670208).

**Comment by rnons** [comment]
[D3 4.0 released](https://github.com/d3/d3/releases/tag/v4.0.0)

**Comment by rreusser** [comment]
Observations:

- I'm sure rollup.js is a wonderful piece of software, but it's appears rather difficult to work into the existing build process because d3 imports only affect the d3 through implied side-effects. 😞  Example: I tried to integrate [d3-selection-multi](https://github.com/d3/d3-selection-multi). See: https://github.com/d3/d3-selection-multi/issues/5 for why this is difficult. The only reasonable answer seemed to be to just copy/paste code from d3 modules into the plotly codebase. A more carefully considered solution pulls of some fancy trick I'm not able to figure out related to ES6 modules and their interaction with browserified code.
- `style({...})` -> `styles({...})`. Ditto attrs. This change is mostly made in the [d3-upgrade](https://github.com/plotly/plotly.js/tree/d3-upgrade) branch.
- selections are no longer subclasses of arrays. This means the `.selectAll(...)[0][0]` indexing trick is now invalid.
- `.data([0])` works fine, but the style through which it's currently achieved breaks since the resulting selection doesn't actually contain the created elements. It's a simply fix. I think I got most of them.
- `d3.round` is gone and seems not to have a replacement. I fudged it with a quick fix, but this might require more careful consideration.

Beyond that, it's not like the logic is all that different. It's a lot of bouncing from one error to the next and translating the code.

**Comment by monfera** [comment]
@rreusser nice summary! Yes, lots of relatively superficial upgrade work even with smaller codebases. Various incompatible changes, e.g. the collapsing of the namespaces, various renamings or both (e.g. `d3.layout.force` -> `d3.forceSimulation`), or what things return (e.g. some palettes were scales but now they're arrays) or what methods need to be used.

Besides the `d3.round` thing, there are subtle changes, such as `d3.range` no longer attempting to control for  numerical inaccuracy ([I saw it from Irene Ros, great presentation](https://iros.github.io/d3-v4-whats-new/#52)).

I wonder if the `selection[0][0]` thing can be replaced with `.node()`. In any case the array access was never part of the API so it's a good idea to avoid it.

Regarding the promise of smaller components, maybe [this new tool can help customize what's included](https://tungs.github.io/browser-roller/d3Roller.html) but I haven't yet tried. The reason it might work is that there are numerous modules not used in the codebase. In any case it's an almost negligibly small fraction of the `plotly.js` codebase.

Based on the current way we use D3 I wouldn't expect significant user-visible improvements from a switch. It can be useful for other reasons, I suppose the community (programmers, bug fixes) move to 4.0 over time. Or if there is a big modularization attempt in `plotly.js` itself, e.g. now lightweight plots as in journals aren't feasible due to the load and initialization time. Some other things also came up earlier, e.g. our use of the palettes (there's duplication between `plotly.js` and the `gl-vis` packages. The force layout in `4.0` is deterministic and *way* faster than currently, see [my three recent blocks](http://bl.ocks.org/monfera) or [these two gifs from Mike Bostock](https://twitter.com/mbostock/status/725008899569422338). 

All in all `D3 4.0` doesn't seem to address a current technical problem but makes some things easier in the future. It's a tough call because most folks can leave old projects in `3.*` and use `4.*` on new ones but `plotly.js` is not a dataviz project.

**Comment by rreusser** [comment]
Haha, but array access is soo easy! Come to think of it, I'm not sure it's actually used in the code. That might just be what I use to debug in the inspector.

**Comment by rreusser** [comment]
Yeah, the other change that maybe had consequences was transitions. The current animation PR falls back to instantaneous data updates when you transition the scale and the data at the same time. The reason seems to be a race condition in which the running transitions (points moving around) don't play well with transform changes (the whole layer moving around). It's difficult to debug because of the way v3 makes it relatively difficult to group transitions (i.e. `transition.select()`; and difficult at least within the plotly.js codebase). v4 lets you group more precisely by simply passing a transition to a new transition.

It's ripe for debugging and refactoring, but the current priority is consistency and correctness, then handling more complicated interactions and corner cases.

**Comment by monfera** [comment]
Good point. Transitions can be a huge thing to upgrade properly (the new is so much better, taking advantage of it may need full rethinking of the transitions). Since you wrote the animations you are the one who knows about the expected `transition` upgrade friction.

**Comment by etpinard** [comment]
Referencing @rreusser 's https://github.com/plotly/plotly.js/pull/946#discussion_r82048108 about our the very common `.data([0])` pattern.

**Comment by rreusser** [comment]
Makes it an extra good reason to factor it out so it's only *one* fix.

**Comment by agnivade** [comment]
Hi there, this looks like a great project ! What is the status of upgrading to d3 v4 ? Has this been put to the backburner for now ?

**Comment by monfera** [comment]
@agnivade thanks! Not quite an answer, just wanted to mention that we have a little bit of use of a couple of D3 4.0 modules where it had a significant performance benefit - the use of the force layout, and its dependencies eg. quadtree in Sankey are 4.0. Force is being used to settle the diagram when moving around nodes: https://codepen.io/monfera/full/VbKrXM/

**Comment by agnivade** [comment]
I see. Thanks, makes sense to me.

**Comment by etpinard** [comment]
FYI d3 v5 should make its official debut soon: https://github.com/d3/d3/releases/tag/v5.0.0

**Comment by mojoaxel** [comment]
Well, it's 2018 and d3 is in v5 by now... Is plotly planning to upgrade at some point?

**Comment by namoshizun** [comment]
Any news on this matter? Is this D3 version bumping still considered?
Have been recently using the R plotly but occasionly had to access the D3 interface. It is a bit cumbersome to use the old D3 v3..

**Comment by nite** [comment]
Any progress on this fellas?

**Comment by etpinard** [comment]
Pasting https://github.com/plotly/plotly.js/issues/3052#issuecomment-424744313 

-----

For completeness, the items d3v3 is missing that v5 has are:

`%Q` - milliseconds since UNIX epoch.

`%s` - seconds since UNIX epoch.

`%u` - Monday-based (ISO 8601) weekday as a decimal number [1,7].

`%V` - ISO 8601 week of the year as a decimal number [01, 53].

`%f` - microseconds as a decimal number [000000, 999999]. Frustratingly, d3 did something different than we did with `%f`, which we added (before it was part of d3) to support more digits and increased flexibility over `%L`. They added it as a fixed 6 digits of precision, whereas we support any precision up to 6 digits (using `%{n}f`) and we truncate trailing zeros. The fixed precision of only 3 (`%L`) or 6 digits is, IMO, a significant limitation (though I can see wanting to keep trailing zeros, we could easily add that) so I guess we'll be keeping this particular difference from D3 long-term...

Perhaps not worth it until we upgrade our d3, but it would be possible for us to add `Q,s,u,V` similarly to how we added `f`.

**Comment by nite** [comment]
d3-format v4+ has new formatting features ... 

eg.
```
d3.format("(0.2s")(-1500);
(2k)
```

**Comment by bishonen** [comment]
if there's no real plans to update to the newer version of d3, is there some recommendation for users to still end up with a ISO8601 weeknum (%V)?

**Comment by kristapskristaps** [comment]
> if there's no real plans to update to the newer version of d3, is there some recommendation for users to still end up with a ISO8601 weeknum (%V)?
Is there any plan to implement this?

**Comment by sebhmg** [comment]
Also especially usefully in the new d3-format: 
> The `~` option trims insignificant trailing zeros
 
Finally, In order to have nice tick positioning and formatting while zooming, I had to use newer d3 (5.15) on the side. I imagine Plotly would just take care of it, once upgraded to a newer version of d3.

Meanwhile, here is my code.
Upon `ploty_relayout` event:
```js
        // use d3 to compute ticks (a newer version than the one used by Plotly)
        const xNiceScale = d3.scaleLinear().domain(xaxisRange).nice();
        const xTickValues = xNiceScale.ticks();
        const xTickFormatter = xNiceScale.tickFormat(xTickValues.length);
        const xTickLabels = xTickValues.map(xTickFormatter);

        const yNiceScale = d3.scaleLinear().domain(yaxisRange).nice();
        const yTickValues = yNiceScale.ticks();
        const yTickFormatter = yNiceScale.tickFormat(yTickValues.length);
        const yTickLabels = yTickValues.map(yTickFormatter);

        const layout = {
            "xaxis.tickmode": "array",
            "xaxis.tickvals": xTickValues,
            "xaxis.ticktext": xTickLabels,
            "yaxis.tickmode": "array",
            "yaxis.tickvals": yTickValues,
            "yaxis.ticktext": yTickLabels
        };
```

**Comment by archmoj** [comment]
> if there's no real plans to update to the newer version of d3, is there some recommendation for users to still end up with a ISO8601 weeknum (%V)?

This one is addressed in #5026 and would be available in `plotly.js v1.55.0`.

**Comment by archmoj** [comment]
:tada: and now d3-v6 is out:
https://observablehq.com/d/f91cccf0cad5e9cb

**Comment by jackparmer** [comment]
This issue has been tagged with `NEEDS SPON$OR`

A [community PR](https://github.com/plotly/plotly.js/pulls?q=is%3Apr+is%3Aopen+label%3A%22type%3A+community%22) for this feature would certainly be welcome, but our experience is deeper features like this are difficult to complete without the Plotly maintainers leading the effort.

**Sponsorship range: $70k-$80k**

What Sponsorship includes:
- Completion of this feature to the Sponsor's satisfaction, in a manner coherent with the rest of the Plotly.js library and API
- Tests for this feature
- Long-term support (continued support of this feature in the latest version of Plotly.js)
- Documentation at [plotly.com/javascript](https://plotly.com/javascript/)
- Possibility of integrating this feature with Plotly Graphing Libraries (Python, R, F#, Julia, MATLAB, etc)
- Possibility of integrating this feature with Dash
- Feature announcement on [community.plotly.com](https://community.plotly.com/tag/announcements) with shout out to Sponsor (or can remain anonymous)
- Gratification of advancing the world's most downloaded, interactive scientific graphing libraries (>50M downloads across supported languages)

Please include the link to this issue when [contacting us](https://plotly.com/get-pricing/) to discuss.

**Comment by jacksongoode** [comment]
Just to be clear... Plotly is based on a version (~3.x) of d3 from 2016... The current version of d3 is now 7.4.5. This seems like a major concern given that it's the foundational package for graphing.

**Comment by gvwilson** [comment]
@emilykl before you move on, could you please create a confluence page with your findings?
